api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,254 @@
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. If multiple headers share the same key, they
+  // must be merged according to the HTTP spec. All header keys must be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool no_body = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a RequestHeaders or ResponseHeaders
+// message.
+message HeaderResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+}
+
+message BodyResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+
+  // Request that the last body chunk delivered be modified.
+  BodyMutation body_mutation = 3;
+}
+
+// This message completely replaces the response.
+// It will immediately return the data specified here to the downstream
+// system and stop further filter processing.
+message ImmediateResponse {
+  type.v3.HttpStatus status = 1 [(validate.rules).message = {required: true}];
+
+  repeated config.core.v3.HeaderValue headers = 2;
+
+  bytes body = 3;
+
+  repeated config.core.v3.HeaderValue trailers = 4;
+}
+
+// Change all or part of the message body by replacing the portion
+// of the last chunk received denoted by "offset" and "length" with "body."
+message BodyMutation {
+  uint32 offset = 1;
+
+  uint32 length = 2;
+
+  bytes body = 3;
+}
+
+// Change the processing mode of the filter. If "duration" is set,
+// then the change is a temporary change that will be un-done after the specified
+// timeout. (If multiple requests arrive after a timer has been set, then the results
```

Comment:
This only applies to the current stream, right?

Reply:
No, it's either permanent (for the Envoy server that received the message) or subject to a duration. This message is only used with the "mode_change" field in the ProcessingResponse.

There is also a "mode_override" field in the ProcessingResponse that changes the mode only for this stream.

Reply:
Per above this seems super complicated. Do we need this in the initial impl? I can also foresee all kinds of race conditions if we can go from async back to sync for example.